grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -23,47 +57,222 @@ class PrepaymentInvoiceService {
+        BigDecimal prepaymentPercent = (orderItem.order.paymentTerm?.prepaymentPercent ?: Constants.DEFAULT_PAYMENT_PERCENT) / 100
+        InvoiceItem prepaymentItem = orderItem.invoiceItems.find { it.isPrepaymentInvoice }
+        InvoiceItem inverseItem = createFromShipmentItem(shipmentItem)
+        // For shipment items we have to check if the ordered quantity was edited after prepayment was generated.
+        // If that was the case we have to check maximum quantity available to inverse
+        Integer quantityInverseable = getQuantityAvailableToInverse(orderItem, prepaymentItem)
+        if (!quantityInverseable || quantityInverseable < 0) {
+            return null
         }
+        Integer quantity = invoiceItem.quantity >= quantityInverseable ? quantityInverseable : invoiceItem.quantity
+        inverseItem.quantity = quantity
+        inverseItem.invoiceItemType = InvoiceItemType.INVERSE
+        inverseItem.unitPrice = prepaymentItem.unitPrice
+        // Multiplied by (-1) to keep inverse items as negative amount
+        inverseItem.amount = quantity * prepaymentItem.unitPrice * prepaymentPercent * (-1)
+        return inverseItem
+    }
 
-        order.activeOrderItems.each { OrderItem orderItem ->
-            InvoiceItem invoiceItem = invoiceService.createFromOrderItem(orderItem)
-            invoice.addToInvoiceItems(invoiceItem)
+    private InvoiceItem createInverseItemForOrderAdjustment(OrderAdjustment orderAdjustment) {
+        InvoiceItem prepaymentItem = orderAdjustment.invoiceItems.find { it.isPrepaymentInvoice }
+        InvoiceItem inverseItem = createFromOrderAdjustment(orderAdjustment)
+        // For order adjustment invoiceItem.quantity is 1 or 0, but for inverse should be for now 1
+        inverseItem.quantity = 1
+        inverseItem.invoiceItemType = InvoiceItemType.INVERSE
+        inverseItem.unitPrice = prepaymentItem.unitPrice
+        // Multiplied by (-1) to keep inverse items as negative amount
+        inverseItem.amount = prepaymentItem.amount * (-1)
+        return inverseItem
+    }
+
+    private Integer getQuantityAvailableToInverse(OrderItem orderItem, InvoiceItem prepaymentItem) {
+        Integer quantityInversed = orderItem.allInvoiceItems.findAll { it.inverseItem }.sum { it.quantity } ?: 0
+        return prepaymentItem.quantity > quantityInversed ? prepaymentItem.quantity - quantityInversed : 0
+    }
+
+    /**
+     * Updates invoice item's quantity and amount and finds related inverse item and updates quantity and amount
+     * */
+    void updateInvoiceItemQuantity(String itemId, Integer quantity) {
+        InvoiceItem invoiceItem = InvoiceItem.get(itemId)
+        if (quantity > invoiceItem.quantity) {
+            throw new IllegalArgumentException("Cannot update quantity invoiced to higher value")
         }
+        if (quantity <= 0) {
+            throw new IllegalArgumentException("Quantity invoiced needs to be higher than 0")
```

Comment:
we should also have a case here for not allowing the user to edit the quantity if quantity is 0 / the shipment item is cancelled. For that case quantity must always be 0.

Reply:
But we don't want to allow cancelling items through this (editing to 0), but I double check that.